src/core/control/tools/StrokeHandler.cpp

```diff
@@ -363,20 +349,39 @@ void StrokeHandler::onButtonDoublePressEvent(const PositionInputData& pos) {
 
 void StrokeHandler::destroySurface() {
     if (surfMask || crMask) {
+        assert(!this->fullRedraw);
         cairo_destroy(crMask);
         cairo_surface_destroy(surfMask);
         surfMask = nullptr;
         crMask = nullptr;
     }
 }
 
-void StrokeHandler::paintDot(const double x, const double y, const double width) const {
-    cairo_set_line_cap(crMask, CAIRO_LINE_CAP_ROUND);
-    cairo_set_operator(crMask, CAIRO_OPERATOR_OVER);
+void StrokeHandler::paintDot(cairo_t* cr, const double x, const double y, const double width) const {
+    cairo_set_line_cap(cr, CAIRO_LINE_CAP_ROUND);
+    cairo_set_line_width(cr, width);
+    cairo_move_to(cr, x, y);
+    cairo_line_to(cr, x, y);
+    cairo_stroke(cr);
+}
+
+void StrokeHandler::createMask() {
+    const double ratio = xournal->getZoom() * static_cast<double>(xournal->getDpiScaleFactor());
+
+    std::unique_ptr<Rectangle<double>> visibleRect(xournal->getVisibleRect(redrawable));
+
+    // We add a padding to limit graphical bugs when scrolling right after completing a stroke
+    const double strokeWidth = this->stroke->getWidth();
+    const int width = static_cast<int>(std::ceil((visibleRect->width + strokeWidth) * ratio));
+    const int height = static_cast<int>(std::ceil((visibleRect->height + strokeWidth) * ratio));
+
+    surfMask = cairo_image_surface_create(CAIRO_FORMAT_A8, width, height);
```

Comment:
I think `surfMask` needs to be destroyed first if it already exists, otherwise there may be a memory leak.

Also, would it be better to use `gdk_window_create_similar_image_surface`?

Reply:
I don't think `surfMask` should ever exist before that.
The entire class was structured before the switch to `GTK3`. Before that, the InputHandlers used to survive a single stroke creation: there was one IntpuHandler of each type for each page.
Now, the class is created upon ButtonPressEvent (see `XojPageView::onButtonPressEvent`). An entire refactor of all the InputHandlers would help clearing all that out: someday, maybe.

I'll add an assert.

Reply:
About `gdk_window_create_similar_image_surface`: at this point, we do not have access to any surface to mimic. We only have a surface in the `draw` loop.